python/GangaDirac/Lib/Backends/DiracBase.py

```diff
@@ -974,8 +957,26 @@ def updateMonitoringInformation(jobs_):
         #logger.debug('Monitor jobs    : ' + repr([j.fqid for j in monitor_jobs]))
         #logger.debug('Requeue jobs    : ' + repr([j.fqid for j in requeue_jobs]))
 
+        # Split all the monitorable jobs into groups based on the
+        # credential used to communicate with DIRAC
+        # TODO Perhaps this would make more sense being done on a more generic level
+        monitor_jobs_by_credential = defaultdict(list)  # type: Dict[ICredentialInfo, List[Job]]
+        for j in monitor_jobs:
+            try:
+                cred_req = j.backend.credential_requirements
+                cred = credential_store[cred_req]
+                if not cred.is_valid():
+                    logger.debug('Required credential %s is not valid', cred)
+                    needed_credentials.add(cred_req)
+                    continue
+                monitor_jobs_by_credential[cred].append(j)
+            except KeyError:
```

Comment:
Can't this just be avoided by using `if element in dict`? The latter is much faster once you start throwing many KeyErrors and I've a dislike for writing code to intentionally throw an exception in a large number of cases it just feels like the code is doing something not quite right.